UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/BMRS_API_Input_JA_7.py

```diff
@@ -0,0 +1,526 @@
+        uri=url,
+        method='GET',
+        headers={'Content-Type': 'application/xml; charset=UTF-8'},
+    )
+    #print('===Response===')
+    #print(pformat(resp))
+    
+    #print('===Content===')
+    return str(pformat(content))
+    #print(type(pformat(content)))
+
+    #print('===Finished===')
+
+
+def float_chars(test):
+    #Check if all characters in the string "test" are valid for a float.
+    valid = "0987654321."
+    for i in test:
+        if i not in valid:
+            return 0
+    return 1
+
+
+def ext_chars(test):
+    #Check if all characters in the string "test" are valid for a float.
+    valid = "0987654321.-_ "
+    for i in test:
+        if i not in valid:
+            return 0
+    return 1
+
+
+def xmlStringProcessToArray(xmlInfo):
+    #Takes the big bulk of text and splits it up into a readable array.
+    xmlInfo = xmlInfo.replace("'\n b'","")
+    xmlInfo = xmlInfo.split("<item>")
+    #Now each element in xmlInfo should contain a potential generator and its output.
+    #A dictionary should be created such that each ID obtains an output.
+    unitOutput = {}
+    for item in xmlInfo:
+        #Check if this is a valid generator, with an output.
+        #Will check if there is a single ID (NGCBMUnitID) and output (quantity). Thus, these terms should exist twice each, with a <> and </> case. 
+        if (2 == len(item.split("<nGCBMUnitID>"))) and (2 == len(item.split("</nGCBMUnitID>"))) and (3 == len(item.split("nGCBMUnitID"))) and (2 == len(item.split("<quantity>"))) and (2 == len(item.split("</quantity>"))) and (3 == len(item.split("quantity"))):
+            #Extract the ID and Quantity values. 
+            ID = re.search("<nGCBMUnitID>(.+?)</nGCBMUnitID>", str(item))
+            quantity = re.search("<quantity>(.+?)</quantity>", str(item))
+            #Check if this worked, if so, just use the values extracted, rather than the current information, which includes if the extraction worked. 
+            if (ID) and (quantity):
+                ID = str(ID.group(1))
+                quantity = str(quantity.group(1))
+                #Check the value contains only valid characters. 
+                if (float_chars(quantity)):
+                    #Add these values to the dictionary. 
+                    unitOutput[ID] = float(quantity)
+                    #print(item)
+                    #print(ID)
+                    #print(quantity)
+    return unitOutput
+
+
+def run_query(Key, Year, Month, Day, Period):
+    #The query function
+    Key = str(Key)
+    Year = str(Year)
+    Month = str(Month)
+    Day = str(Day)
+    Period = str(Period)
+    querystring = 'https://api.bmreports.com/BMRS/B1610/v2?APIKey='+Key+'&SettlementDate='+Year+'-'+Month+'-'+Day+'&Period='+Period+'&NGCBMUnitID=&ServiceType=xml'
+    
+    #Query All
+    #post_elexon_basic(url='https://api.bmreports.com/BMRS/B1610/v2?APIKey=iwx6raw9m7nqq0f&SettlementDate=2021-01-01&Period=1&NGCBMUnitID=&ServiceType=xml',) #All 
```

Comment:
Please remove the API key value from lines like this (there are a few in this file). I would not remove the lines altogether - they might still be useful.